plugins/pkg/file/fs.go

```diff
@@ -29,110 +27,73 @@ var (
 )
 
 type File struct {
-	lock sync.RWMutex
-
-	Name  string
-	mtime time.Time
-}
-
-func (f *File) Mtime() time.Time {
-	f.lock.RLock()
-	defer f.lock.RUnlock()
-	// the returned time.Time should be readonly
-	return f.mtime
-}
-
-func (f *File) SetMtime(t time.Time) {
-	f.lock.Lock()
-	f.mtime = t
-	f.lock.Unlock()
+	Name string
 }
 
-type fs struct {
-	cache *ttlcache.Cache[string, os.FileInfo]
+type Watcher struct {
+	watcher *fsnotify.Watcher
+	files   map[string]bool
+	mu      sync.Mutex
+	done    chan struct{}
 }
 
-func newFS(ttl time.Duration) *fs {
-	loader := ttlcache.LoaderFunc[string, os.FileInfo](
-		func(c *ttlcache.Cache[string, os.FileInfo], key string) *ttlcache.Item[string, os.FileInfo] {
-			info, err := os.Stat(key)
-			if err != nil {
-				logger.Error(err, "reload file info to cache", "file", key)
-				return nil
-			}
-			item := c.Set(key, info, ttlcache.DefaultTTL)
-			logger.Info("update file mtime", "file", key, "mtime", item.Value().ModTime())
-			return item
-		},
-	)
-	cache := ttlcache.New(
-		ttlcache.WithTTL[string, os.FileInfo](ttl),
-		ttlcache.WithLoader[string, os.FileInfo](loader),
-	)
-	go cache.Start()
-
-	return &fs{
-		cache: cache,
+func NewWatcher() (*Watcher, error) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
 	}
+	return &Watcher{
+		watcher: w,
+		files:   make(map[string]bool),
+		done:    make(chan struct{}),
+	}, nil
 }
 
-var (
-	// TODO: rewrite it to use inotify
-	defaultFs = newFS(10 * time.Second)
-)
-
-func IsChanged(files ...*File) bool {
+func (w *Watcher) AddFile(files ...*File) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
 	for _, file := range files {
-		changed := defaultFs.isChanged(file)
-		if changed {
-			return true
+		if _, exists := w.files[file.Name]; !exists {
```

Comment:
As explained in https://github.com/mosn/htnn/pull/614#discussion_r1658077214, we should watch the directory instead. And be careful that multiple files may use the same directory.

Reply:
why we should be careful that multiple files may use the same directory?